net/tshttpproxy/tshttpproxy_synology.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build linux
+// +build linux
+
+package tshttpproxy
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+
+	"tailscale.com/util/lineread"
+	"tailscale.com/version/distro"
+)
+
+func init() {
+	if distro.Get() == distro.Synology {
+		sysProxyFromEnv = synologyProxyFromConfigCached
+	}
+}
+
+var cachedProxy struct {
+	sync.Mutex
+	val        *url.URL
+	lastUpdate time.Time
+}
+
+func synologyProxyFromConfigCached(req *http.Request) (*url.URL, error) {
+	cachedProxy.Lock()
+	defer cachedProxy.Unlock()
+
+	if time.Since(cachedProxy.lastUpdate) < time.Minute {
+		return cachedProxy.val, nil
+	}
+
+	defer func() {
+		cachedProxy.lastUpdate = time.Now()
+	}()
+
+	val, err := synologyProxyFromConfig(req)
+	cachedProxy.val = val
+	return val, err
+}
+
+func synologyProxyFromConfig(req *http.Request) (*url.URL, error) {
+	if req.URL == nil {
+		return nil, nil
+	}
+
+	r, err := openSynologyProxyConf()
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	defer r.Close()
+
+	return parseSynologyConfig(r)
+}
+
+var openSynologyProxyConf = func() (io.ReadCloser, error) {
+	return os.Open("/etc/proxy.conf")
+}
+
+func parseSynologyConfig(r io.Reader) (*url.URL, error) {
+	cfg := map[string]string{}
+
+	if err := lineread.Reader(r, func(line []byte) error {
+		// accept and skip over empty lines
+		line = bytes.TrimSpace(line)
+		if len(line) == 0 {
+			return nil
+		}
+
+		parts := bytes.SplitN(line, []byte("="), 2)
+		if len(parts) != 2 {
+			return fmt.Errorf("missing \"=\" in proxy.conf line: %q", line)
+		}
+		key, value := parts[0], parts[1]
+		cfg[string(key)] = string(value)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	if cfg["proxy_enabled"] != "yes" {
+		return nil, nil
+	}
+
+	var proxyURL url.URL
```

Comment:
proxyURL := new(url.URL)

to keep it as a pointer throughout, since that's what you're returning.

avoids accidental value copies in the future